### PR TITLE
[DON'T MERGE]: Filtering tenant ID for migration

### PR DESCRIPTION
**Pull request by realsystem**

### CODE REVIEW

File: cloudferrylib/os/compute/nova_compute.py [hunk 1]
```diff
@@ -168,7 +180,10 @@ def read_info(self, target='instances', **kwargs):
 
         for instance in self.get_instances_list(search_opts=search_opts):
             if instance.status in ALLOWED_VM_STATUSES:
-                info['instances'][instance.id] = self.convert(instance,
+                if (self.cloud.position == 'dst') or \
+                    not self.filter_tenant_id or \
```
Comment (realsystem) [inline cloudferrylib/os/compute/nova_compute.py]: tab should be here

File: cloudferrylib/os/compute/nova_compute.py [hunk 2]
```diff
@@ -168,9 +183,12 @@ def read_info(self, target='instances', **kwargs):
 
         for instance in self.get_instances_list(search_opts=search_opts):
             if instance.status in ALLOWED_VM_STATUSES:
-                info['instances'][instance.id] = self.convert(instance,
-                                                              self.config,
-                                                              self.cloud)
+                if (self.cloud.position == 'dst') or \
```
Comment (mirrorcoder) [inline cloudferrylib/os/compute/nova_compute.py]: Зачем это условие здесь?

File: cloudferrylib/os/identity/keystone.py [hunk 3]
```diff
@@ -130,13 +131,22 @@ def read_info(self, **kwargs):
         service_tenant_id = \
             self.get_tenant_id_by_name(self.config.cloud.service_tenant)
 
+        admin_tenant_id = \
+            self.get_tenant_id_by_name(self.config.cloud.admin_tenant)
+
+        if kwargs.get('tenant_id'):
+            self.filter_tenant_id = kwargs['tenant_id']
+
+        tmp_list = [admin_tenant_id, service_tenant_id, self.filter_tenant_id]
```
Comment (mirrorcoder) [inline cloudferrylib/os/identity/keystone.py]: What is the tmp_list?

Comment (realsystem) [inline cloudferrylib/os/identity/keystone.py]: Just a list of items to compare below

File: cloudferrylib/os/compute/nova_compute.py (hunk 2 above)
Comment (realsystem) [inline cloudferrylib/os/compute/nova_compute.py]: Just extra code(old version), to remove

File: cloudferrylib/os/identity/keystone.py [hunk 4]
```diff
@@ -130,13 +131,22 @@ def read_info(self, **kwargs):
...
+            self.get_tenant_id_by_name(self.config.cloud.admin_tenant)
+
+        if kwargs.get('tenant_id'):
+            self.filter_tenant_id = kwargs['tenant_id'][0]
+
+        tmp_list = [admin_tenant_id, service_tenant_id, self.filter_tenant_id]
         for tenant in self.get_tenants_list():
-            if tenant.id != service_tenant_id:
+            if (self.filter_tenant_id and tenant.id in tmp_list) or \
+                    not self.filter_tenant_id:
```
Comment (mirrorcoder) [inline cloudferrylib/os/identity/keystone.py]: Before this, service_tenant_id parameter was used to ignore the tenant service, and now there is.

**Comment by roman-verchikov** [comment]
Let's also update README.md with single tenant migration documentation

File: cfglib.py [hunk 5]
```diff
@@ -136,7 +140,10 @@
                      'OpenStack does not support user quotas (e.g. Grizzly)'),
     cfg.StrOpt('incloud_live_migration', default='nova',
                help='Live migration type used for in-cloud live migration. '
-                    'Possible values: "nova", "cobalt".')
+                    'Possible values: "nova", "cobalt".'),
+    cfg.StrOpt('filter_tenant_id', default='-',
```
Comment (miarmak) [inline cfglib.py]: Do we use it somewhere?

File: cloudferrylib/os/image/glance_image.py [hunk 6]
```diff
@@ -90,10 +91,29 @@ def get_client(self):
...
+            for img in get_img_list(filters=filters):
+                image_list.append(img)
+            filters = {'is_public': None}
+            for img in get_img_list(filters=filters):
+                # getting images if tenant is member
+                for entry in self.glance_client.image_members.list(image=img):
+                    if entry.member_id == self.filter_tenant_id:
+                        image_list.append(img)
+                # getting public images
+                for k, v in img.to_dict().items():
```
Comment (miarmak) [inline cloudferrylib/os/image/glance_image.py]: Doesn't image object have attribute `is_public`?

File: configs/filter.yaml [hunk 7]
```diff
@@ -1,3 +1,7 @@
+#tenants:
+#    tenant_id:
+#        - <tenant_id1>
+#        - <tenant_id2>
```
Comment (miarmak) [inline configs/filter.yaml]: I guess we should write somewhere, that with this patch we can filter only by 1 tenant... And I guess with those one, that has been specified first. Isn't is?

Comment (realsystem) [inline configs/filter.yaml]: Will be done after https://github.com/MirantisWorkloadMobility/CloudFerry/pull/303

File: cfglib.py (hunk 5 above)
Comment (realsystem) [inline cfglib.py]: no, to be removed

File: cloudferrylib/os/image/glance_image.py (hunk 6 above)
Comment (realsystem) [inline cloudferrylib/os/image/glance_image.py]: good catch

File: cloudferrylib/os/identity/keystone.py [hunk 8]
```diff
@@ -416,12 +432,11 @@ def _get_user_passwords(self):
 
         return info
 
-    def _get_user_tenants_roles(self):
+    def _get_user_tenants_roles(self, tenant_list=[], user_list=[]):
```
Comment (miarmak) [inline cloudferrylib/os/identity/keystone.py]: I would not recommend use mutable default arguments.
http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments

File: cloudferrylib/os/network/neutron.py [hunk 9]
```diff
@@ -52,10 +53,33 @@ def read_info(self, **kwargs):
...
+            for net in admin_nets:
+                if net['shared']:
+                    nets.append(net)
+            # getting all admin subnets
+            admin_subnets = self.get_subnets(admin_tenant_id)
+            # getting subnets for shared and tenant nets
+            for subnet in admin_subnets:
+                for net in nets:
+                    if subnet['network_name'] == net['name']:
+                        subnets.append(subnet)
```
Comment (roman-verchikov) [inline cloudferrylib/os/network/neutron.py]: network names may be identical, we should only rely on UUIDs

File: cloudferrylib/os/network/neutron.py [hunk 10]
```diff
@@ -52,10 +53,33 @@ def read_info(self, **kwargs):
...
+        else:
+            tenant_id = ''
+
+        nets = self.get_networks(tenant_id)
+        subnets = self.get_subnets(tenant_id)
+
+        if self.filter_tenant_id:
+            # getting all admin nets
+            admin_tenant_id = self.identity_client.get_tenant_id_by_name(
+                self.config.cloud.admin_tenant)
```
Comment (roman-verchikov) [inline cloudferrylib/os/network/neutron.py]: `admin_tenant` option already exists in config, please reuse `tenant`

File: cloudferrylib/os/network/neutron.py [hunk 11]
```diff
@@ -52,10 +53,33 @@ def read_info(self, **kwargs):
...
+            admin_nets = self.get_networks(admin_tenant_id)
+            # getting admin shared nets
+            for net in admin_nets:
+                if net['shared']:
+                    nets.append(net)
```
Comment (roman-verchikov) [inline cloudferrylib/os/network/neutron.py]: See if shared net list can be retrieved with single neutron API call

**Comment by roman-verchikov** [comment]
```
2 floating IPs did not migrate to destination: set([u'172.18.10.2', u'172.18.10.3'])
```
floating IP tests must be updated accordingly

File: cloudferrylib/os/network/neutron.py [hunk 12]
```diff
@@ -1152,7 +1166,7 @@ def upload_floatingips(self, networks, src_floats):
                 tenant = self.identity_client.keystone_client.tenants.find(
                     name=fip['tenant_name'])
 
-                if self._filtered_tenant(tenant.id):
+                if self.filter_tenant_id != tenant.id:
```
Comment (roman-verchikov) [inline cloudferrylib/os/network/neutron.py]: Should be changed to
```
if self.filter_tenant_id is not None and self.filter_tenant_id != tenant.id:
```

File: cloudferrylib/os/network/neutron.py (hunk 11 above)
Comment (roman-verchikov) [inline cloudferrylib/os/network/neutron.py]: Can be slightly improved with
```
shared_nets = self.neutron_client.list_networks(shared=True)
external_nets = self.neutron_client.list_networks(**{'router:external': True})
```

**Pull request merged**